lib/rspec/core/bisect/example_minimizer.rb

```diff
@@ -18,20 +16,52 @@ def initialize(runner, reporter)
         def find_minimal_repro
           prep
 
-          self.remaining_ids = non_failing_example_ids
+          self.remaining_ids, duration = track_duration do
+            bisect_over(non_failing_example_ids)
+          end
 
-          each_bisect_round do |subsets|
-            ids_to_ignore = subsets.find do |ids|
-              get_expected_failures_for?(remaining_ids - ids)
-            end
+          notify(:bisect_complete, :round => 10, :duration => duration,
+                                   :original_non_failing_count => non_failing_example_ids.size,
+                                   :remaining_count => remaining_ids.size)
+
+          remaining_ids + failed_example_ids
+        end
 
-            next :done unless ids_to_ignore
+        def bisect_over(candidate_ids)
+          return [] if get_expected_failures_for?([])
 
-            self.remaining_ids -= ids_to_ignore
-            notify(:bisect_ignoring_ids, :ids_to_ignore => ids_to_ignore, :remaining_ids => remaining_ids)
+          bisect_over_recursive(candidate_ids, [])
+        end
+
+        def bisect_over_recursive(candidate_ids, fixed)
+          return candidate_ids if candidate_ids.length == 1
+          slice_size = (candidate_ids.length / 2.0).ceil
+          slices = candidate_ids.each_slice(slice_size).to_a
+
+          notify(
+            :bisect_round_started,
+            :round => 10,
+            :subset_size => slice_size,
+            :remaining_count => candidate_ids.size
+          )
+
+          ids_to_ignore = slices.find do |ids|
+            get_expected_failures_for?(candidate_ids - ids + fixed)
           end
 
-          currently_needed_ids
+          if ids_to_ignore
+            self.remaining_ids = candidate_ids - ids_to_ignore
+            notify(
+              :bisect_ignoring_ids,
+              :ids_to_ignore => ids_to_ignore,
+              :remaining_ids => remaining_ids
+            )
+            bisect_over_recursive(remaining_ids, fixed)
+          else
+            slices.permutation(2).map do |slice, other_slice|
+              bisect_over_recursive(slice, fixed + other_slice)
+            end.flatten(1)
```

Comment:
I had to think for a while about this to confirm that there's always only two permutations here.  IMO, this would be easier to understand if you explicitly spelled out the two permutations rather than using `permutation(2).map { }.flatten(1)`.  This also works out well to let us name the two sides `lhs` and `rhs` which helped me understand the algorithm and your diagram:

``` ruby
lhs, rhs = candidate_ids.each_slice(slice_size).to_a

# ...

ids_to_ignore = [lhs, rhs].find do |ids|
  get_expected_failures_for?(candidate_ids - ids + fixed)
end

if ids_to_ignore
  # ...
else
  bisect_over_recursive(lhs, fixed + rhs) +
  bisect_over_recursive(rhs, fixed + lhs)
end
```

This simpler form may also help with this:

> At the moment because the code is clearer when the recursive steps are independent. It's certainly an optimisation that could be applied, but my (brief) attempt to do so resulted in a bit of a mess, so I thought I'd submit the PR with the more understandable code. I'll have another look at this.


...since they are no longer generically run through a `map` block the 2nd recursive call can more easily be updated to account for what was learned from the 1st recursive call.